### PR TITLE
IO refactoring

### PR DESCRIPTION
**Pull request by GWeindel**
I'm working on a complete refactoring of IO. The main ojective is to have better code but also to be able to read and preprocess any BIDS dataset.

So far this version works for BIDS (tests broken for now), not mne.raw nor mne.epochs yet.
It is quite easy to analyze a BIDS dataset, e.g. this is the only code I need to analyze ERP core flanker task after downloading it:

```python
import hmp

sfreq=250

# Which triggers to center the epochs on
centering_id = {
       "compatible/left" : 11,
	   "compatible/right" : 12,
	   "incompatible/left" : 21,
	   "incompatible/right" : 22,
}

# Which triggers indicate the response/the end of the duration to be modelled
response_id = {
	   "left/compatible/correct":111,
	   "left/incompatible/correct":121,
	   "right/compatible/correct":212,
	   "right/incompatible/correct":222
}

# BIDS argument, root and selection (datatype, session, task, ...)
bids_kwargs = dict(
    root = "../../data/Flanker", #Mandatory
    datatypes=['eeg'],#Mandatory
    tasks=['LRP'],#Mandatory
    subjects=['001','002'],
)

# Epoching arguments, anything that can be used by mne.Epoch
epoching_kwargs = dict(
    tmin = -0.2,
    tmax = 2,
)

# Preprocessing arguments, can have low_pass, high_pass, sfreq, etc.
preprocessing_kwargs = dict(
    sfreq = sfreq,
)

# Reading the data
epoch_data, info = hmp.io.read_bids_raw(
    bids_kwargs=bids_kwargs,
    epoching_kwargs=epoching_kwargs,
    preprocessing_kwargs=preprocessing_kwargs,
    centering_id = centering_id,
    response_id = response_id,
    # preprocessing_fn=preprocessing, # users can now also use their own preprocessing functions
    montage='biosemi64',
    cpus=-2, #Can also read data in parallel
)

hmp_data = hmp.preprocessors.ProjPCA(epoch_data, interval_id='rt',
                                    min_duration=0.2, # minimum RT of 200ms
                                    n_comp=10, #need for speed
)


model = hmp.models.CumulativeMethod()
lkh, estimates = model.fit_transform(hmp_data)
``` 

The only thing that changes for downstream functions is that the `participant` dimension has to be replaced by `recording`:

<div class="lm-Widget jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-active jp-mod-selected" aria-label="Code Cell Content with Output" data-windowed-list-index="1" tabindex="0" style="contain-intrinsic-size: auto 38.6px; content-visibility: visible; contain: style;"><div class="lm-Widget lm-Panel jp-Cell-outputWrapper"><div class="lm-Widget jp-OutputArea jp-Cell-outputArea" style=""><div class="lm-Widget lm-Panel jp-OutputArea-child"><div class="lm-Widget jp-RenderedText jp-mod-trusted jp-OutputArea-output" data-mime-type="application/vnd.jupyter.stdout"><pre>&lt;xarray.Dataset&gt; Size: 53MB
Dimensions:     (recording: 2, epoch: 400, channel: 30, sample: 551)
Coordinates:
  * recording   (recording) &lt;U16 128B 'sub-001_task-LRP' 'sub-002_task-LRP'
    subject     (recording) &lt;U3 24B '001' '002'
    task        (recording) &lt;U3 24B 'LRP' 'LRP'
  * epoch       (epoch) int64 3kB 0 1 2 3 4 5 6 ... 393 394 395 396 397 398 399
    event_name  (recording, epoch) object 6kB 'stimulus/compatible/left' ... ...
    rt          (recording, epoch) float64 6kB 0.552 0.432 0.508 ... 0.512 0.628
    response    (recording, epoch) object 6kB 'left/compatible/correct' ... '...
  * channel     (channel) &lt;U3 360B 'Fp1' 'F3' 'F7' 'FC3' ... 'PO8' 'PO4' 'O2'
  * sample      (sample) int64 4kB -50 -49 -48 -47 -46 ... 496 497 498 499 500
Data variables:
    data        (recording, epoch, channel, sample) float32 53MB 3.228e-06 .....
Attributes: (12/17)
    root:                  ../../data/Flanker
    datatypes:             ['eeg']
    tasks:                 ['LRP']
    subjects:              ['001', '002']
    tmin:                  -0.2
    tmax:                  2
    ...                    ...
    reject_by_annotation:  False
    sfreq:                 250
    high_pass:             0.01
    low_pass:              40
    reference:             average
    pick_channels:         eeg
</pre></div></div><div class="jp-OutputArea-promptOverlay" title="Collapse Output" style="height: 629px;"><svg xmlns="http://www.w3.org/2000/svg" width="16" viewBox="0 0 8.5 10.5" data-icon="ui-components:collapse" data-icon-id="00dffbaf-2a9e-434c-9622-87ab6fe9f06a" aria-hidden="true"></svg></div></div></div><div class="lm-Widget jp-CellFooter jp-Cell-footer"></div></div><div class="lm-Widget jp-Cell jp-CodeCell jp-Notebook-cell" aria-label="Code Cell Content with Output" data-windowed-list-index="2" tabindex="-1" style="contain-intrinsic-size: auto 38.6px; content-visibility: visible; contain: style;"><div class="lm-Widget jp-CellHeader jp-Cell-header"></div><div class="lm-Widget lm-Panel jp-Cell-inputWrapper"><div class="lm-Widget jp-Collapser jp-InputCollapser jp-Cell-inputCollapser"><div class="jp-Collapser-child"></div></div><div class="lm-Widget jp-InputArea jp-Cell-inputArea" style="--neutral-layer-card-container: #333333; --neutral-layer-floating: #4a4a4a; --neutral-layer-1: #3b3b3b; --neutral-layer-2: #333333; --neutral-layer-3: #2b2b2b; --neutral-layer-4: #242424; --fill-color: #3b3b3b; --accent-fill-rest: #3195e4; --accent-fill-hover: #36a0f4; --accent-fill-active: #2b87cf; --accent-fill-focus: #3195e4; --accent-foreground-rest: #4eabf5; --accent-foreground-hover: #6ab9f7; --accent-foreground-active: #3aa2f4; --accent-foreground-focus: #4eabf5; --neutral-fill-rest: #4c4c4c; --neutral-fill-hover: #545454; --neutral-fill-active: #474747; --neutral-fill-focus: #3b3b3b; --neutral-fill-input-rest: #3b3b3b; --neutral-fill-input-hover: #3b3b3b; --neutral-fill-input-active: #3b3b3b; --neutral-fill-input-focus: #3b3b3b; --neutral-fill-stealth-rest: #3b3b3b; --neutral-fill-stealth-hover: #474747; --neutral-fill-stealth-active: #424242; --neutral-fill-stealth-focus: #3b3b3b; --neutral-fill-strong-rest: #a7a7a7; --neutral-fill-strong-hover: #bbbbbb; --neutral-fill-strong-active: #9a9a9a; --neutral-fill-strong-focus: #a7a7a7; --neutral-fill-layer-rest: #424242; --focus-stroke-outer: #929292; --focus-stroke-inner: #0d3e65; --neutral-foreground-hint: #a7a7a7; --neutral-foreground-rest: #ffffff; --neutral-stroke-rest: #7a7a7a; --neutral-stroke-hover: #a1a1a1; --neutral-stroke-active: #636363; --neutral-stroke-focus: #7a7a7a; --neutral-stroke-divider-rest: #4f4f4f; --error-fill-rest: #e34e43; --error-fill-hover: #f5564b; --error-fill-active: #cd453b; --error-fill-focus: #e34e43; --error-foreground-rest: #f8837b; --error-foreground-hover: #f99a93; --error-foreground-active: #f7746b; --error-foreground-focus: #f8837b; --clear-button-hover: #616161; --clear-button-active: #5c5c5c; --tree-item-expand-collapse-hover: #545454; --tree-item-expand-collapse-selected-hover: #595959;"><div class="lm-Widget jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline"><div class="cm-editor ͼ1 ͼ2 ͼ4 ͼo" style=""><div class="cm-announced" aria-live="polite"></div><div tabindex="-1" class="cm-scroller"><div style="tab-size: 4;" spellcheck="false" autocorrect="off" autocapitalize="off" writingsuggestions="false" translate="no" contenteditable="true" class="cm-content" role="textbox" aria-multiline="true" tabindex="-1" data-language="python"><div class="cm-line">info</div></div><div class="cm-layer cm-layer-above cm-cursorLayer" aria-hidden="true" style="z-index: 150; animation-duration: 1200ms; animation-name: cm-blink;"></div><div class="cm-layer cm-selectionLayer" aria-hidden="true" style="z-index: -2;"></div></div></div></div></div></div><div class="lm-Widget lm-Panel jp-Cell-outputWrapper"><div class="lm-Widget jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser"><div class="jp-Collapser-child"></div></div><div class="lm-Widget jp-OutputArea jp-Cell-outputArea" style=""><div class="lm-Widget lm-Panel jp-OutputArea-child jp-OutputArea-executeResult"><div class="lm-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output" data-mime-type="text/html">

But then we can easily analyze different sessions/tasks/run, etc.

This might have a small consequence on the LOOCV PR @jelmerborst but then again if we want to do single case studies (same participant multiple sessions) the LOOCV could also be preformed at a different level (e.g. sessions in this case)

I still have quite some work on testing and reading other formats (#285), I expect to be finished somewhere next week

### CODE REVIEW

**Comment by jelmerborst** [comment]
OK, looks good. We do have `subject`, right? I can build the LOOCV in such a way that it either uses `subject` (default) or `recording`.

**Comment by GWeindel** [comment]
> OK, looks good. We do have `subject`, right? I can build the LOOCV in such a way that it either uses `subject` (default) or `recording`.

Yes, `subject` is part of the BIDS naming convention so it will always be there.
